problem1/problem1.cpp

```diff
@@ -0,0 +1,37 @@
+class Foo {
+private:
+    mutex m;
+    condition_variable cv;
+    int turn;
+    
+public:
+    Foo() : turn(0) { }
+
+    void first(function<void()> printFirst) {
+        turnWrapper(printFirst, 0);
+    }
+
+    void second(function<void()> printSecond) {
+        turnWrapper(printSecond, 1);
+    }
+
+    void third(function<void()> printThird) {
+        turnWrapper(printThird, 2);
+    }
+
+    void turnWrapper(function<void()> whichPrint, int whichTurn) {
+        unique_lock<mutex> lock(m);
+        cv.wait(lock, [this, whichTurn]{return turn == whichTurn;});
+
+        try {
+            whichPrint();
+            fflush(stdout); // Woe to he who uses std::cout in a multithreaded environment.
```

Comment:
if your task was to do print("first") yourself then flushing would be appropriate. 
However this problem should be interpreted as ordering of invocations of third-party functions, rather than actual printing. The actual way how this function prints stuff doesn't have to be on stdout (and can even change over time), so flushing stdout here won't be layer-appropriate.

Although, if you are solving this on the whiteboard in front of interviewer - asking interviewer if original function does flushing and if you need to take care about this would be a appropriate demonstration of skills.